cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetTypeService.java

```diff
@@ -0,0 +1,416 @@
+   * @throws NotFoundException if the namespace in which the module is being added is not found
+   * @throws IOException if there are issues while performing I/O like creating temporary directories, moving/unpacking
+   *                      module jar files
+   * @throws DatasetModuleConflictException if #forceUpdate is {@code false}, and there are conflicts with other modules
+   */
+  BodyConsumer addModule(final DatasetModuleId datasetModuleId, final String className,
+                         final boolean forceUpdate) throws Exception {
+    NamespaceId namespaceId = datasetModuleId.getParent();
+    final Principal principal = authenticationContext.getPrincipal();
+    // enforce that the principal has WRITE access on the namespace
+    authorizationEnforcer.enforce(namespaceId, principal, Action.WRITE);
+    if (NamespaceId.SYSTEM.equals(namespaceId)) {
+      throw new UnauthorizedException(String.format("Cannot add module '%s' to '%s' namespace.",
+                                                    datasetModuleId.getModule(), datasetModuleId.getNamespace()));
+    }
+    ensureNamespaceExists(namespaceId);
+
+    // It is now determined that a new dataset module will be deployed. First grant privileges, then deploy the module.
+    // If creation fails, revoke the granted privileges. This ensures that just like delete, there may be orphaned
+    // privileges in rare scenarios, but there can never be orphaned datasets.
+    // If the module previously existed and was deleted, but revoking privileges somehow failed, there may be orphaned
+    // privileges for the module. Revoke them first, so no users unintentionally get privileges on the dataset.
+    revokeAllPrivilegesOnModule(datasetModuleId);
+    grantAllPrivilegesOnModule(datasetModuleId, principal);
+    try {
+      return createModuleConsumer(datasetModuleId, className, forceUpdate, principal);
+    } catch (Exception e) {
+      revokeAllPrivilegesOnModule(datasetModuleId);
+      throw e;
+    }
+  }
+
+  /**
+   * Deletes the specified {@link DatasetModuleId}
+   */
+  void delete(DatasetModuleId datasetModuleId) throws Exception {
+    NamespaceId namespaceId = datasetModuleId.getParent();
+    if (NamespaceId.SYSTEM.equals(namespaceId)) {
+      throw new UnauthorizedException(String.format("Cannot delete module '%s' from '%s' namespace.",
+                                                    datasetModuleId.getModule(), datasetModuleId.getNamespace()));
+    }
+    ensureNamespaceExists(namespaceId);
+
+    Id.DatasetModule module = datasetModuleId.toId();
+    DatasetModuleMeta moduleMeta = typeManager.getModule(module);
+    if (moduleMeta == null) {
+      throw new DatasetModuleNotFoundException(module);
+    }
+
+    Principal principal = authenticationContext.getPrincipal();
+    authorizationEnforcer.enforce(datasetModuleId, principal, Action.ADMIN);
+
+    try {
+      typeManager.deleteModule(module);
```

Comment:
file a JIRA to track cleanup of `TypeManager#deleteModule` method? It returns a boolean, but the return value is unused. This can be confusing.